core/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -236,10 +245,13 @@ void onFirstPhaseResult(final int shardIndex, @Nullable ShardRouting shard, @Nul
             final ShardRouting nextShard = shardIt.nextOrNull();
             final boolean lastShard = nextShard == null;
             // trace log this exception
-            if (logger.isTraceEnabled()) {
-                logger.trace("{}: Failed to execute [{}] lastShard [{}]", e, shard != null ? shard.shortSummary() : shardIt.shardId(),
-                    request, lastShard);
-            }
+            logger.trace(
+                    () -> new ParameterizedMessage(
```

Comment:
Does this allocate the closure every time or is the JVM able to prevent the allocation if trace isn't enabled?